FWI-3600 First implementation of the realtime-reporter with Polaris support

jslivka opened the pull request:
This PR introduces a `realtime-reporter` plugin for insights:

* Supports Polaris reports only at this time
* Utilizes [SharedInformer](https://pkg.go.dev/k8s.io/client-go@v0.29.0/tools/cache#SharedInformer) from `k8s.io/client-go` to send add/update/delete events
* add and update events pass their manifests to individual handlers for each report type
* reports are wrapped in an event envelope and sent to the insights incremental upload endpoint

## Testing

```sh
# credential setup
export FAIRWINDS_TOKEN=$TOKEN

go run main.go \
    --organization acme-co \
    --cluster kind \
    --host http://192.168.1.27:3000 \
    --config examples/realtime-reporter.yaml \
    --polaris-enabled \
    --polaris-config $HOME/git/fairwinds/polaris/examples/config.yaml
```

```sh
# no watcher provided
go run main.go \
    --organization acme-co \
    --cluster kind \
    --host http://192.168.1.27:3000 \
    --config examples/realtime-reporter.yaml
INFO[0000] Using config file:examples/realtime-reporter.yaml
false

FATA[0000] Error creating new watcher: no valid handler has been specified
exit status 1
```

```sh
# watcher provided with no config specified
go run main.go \
    --organization acme-co \
    --cluster kind \
    --host http://192.168.1.27:3000 \
    --config examples/realtime-reporter.yaml \
    --polaris-enabled
INFO[0000] Using config file:examples/realtime-reporter.yaml
true

FATA[0000] Error creating new watcher: polaris configuration file path must be provided when Polaris is enabled
exit status 1
```

```sh
# watcher and config specified
go run main.go \
    --organization acme-co \
    --cluster kind \
    --host http://192.168.1.27:3000 \
    --config examples/realtime-reporter.yaml \
    --polaris-enabled \
    --polaris-config $HOME/git/fairwinds/polaris/examples/config.yaml
# success
```

The example polaris configuration used in testing can be found in the main project repo: https://github.com/FairwindsOps/polaris/blob/master/examples/config.yaml

## Checklist
* [x] I have signed the CLA
* [x] I have updated/added any relevant documentation

## Description
### What's the goal of this PR?

Support event-based reporting for insights

### What changes did you make?

Added the first version of the `realtime-reporter` plugin for insights-agent

### What alternative solution should we consider, if any?

vitorvezani commented on realtime-reporter/examples/realtime-reporter.yaml:
```diff
@@ -0,0 +1,14 @@
+  - v1/persistentvolumeclaims
+
+namespaces:
+  - all
```
What are the odds of some customer using `all` as an actual namespace?

jslivka replied:
I'm not sure where this is explicit in https://github.com/kubernetes/client-go/blob/v0.29.1/dynamic/dynamicinformer/informer.go#L35, but providing `all` as a namespace works as intended, as it indicates that we want to watch all namespaces in the cluster. If you change this value to `default` or something else, it will be passed along when creating the dynamic informer and only report events occurring in the provided namespace(s)

vitorvezani commented on realtime-reporter/pkg/handlers/polaris.go:
```diff
@@ -0,0 +1,162 @@
+func PolarisHandler(resourceType string) cache.ResourceEventHandlerFuncs {
+
+	var handler cache.ResourceEventHandlerFuncs
+	handler.AddFunc = func(obj interface{}) {
```
Please, re-check error handling in this function:
- do you really want to log and continue?
- we usually use lower-case first letter in error messages

jslivka replied:
I suppose if we can't process the event at this point it would be reason enough to bail out--we can go with this and see if it helps catch any issues early

vitorvezani replied:
Yea, usually bailing-out is what we usually want, if the process that failed is required downstream... which most of the times is the case

jslivka replied:
Will get these changes pushed shortly

vitorvezani commented on realtime-reporter/pkg/handlers/polaris.go:
```diff
@@ -0,0 +1,162 @@
+		}
+	}
+	handler.UpdateFunc = func(old, new interface{}) {
+		timestamp := getTimestampUnixNanos()
```
ditto above

vitorvezani commented on realtime-reporter/pkg/handlers/polaris.go:
```diff
@@ -0,0 +1,162 @@
+			}
+		}
+	}
+	handler.DeleteFunc = func(obj interface{}) {
```
ditto above

vitorvezani commented on realtime-reporter/pkg/handlers/polaris.go:
```diff
@@ -0,0 +1,162 @@
+	Contents []byte
+}
+
+func polarisReportInfoToMap(report models.ReportInfo) map[string]interface{} {
```
make so this function returns an error instead of `panic(err)`

vitorvezani reviewed (changes requested):
Nice work John, I noticed some places whereas error handling could be improved and the avoidance of panic for resilience

jslivka commented:
I've addressed your comments @vitorvezani, everything is still working as intended. This will need more stress-testing going forward but I hope the changes in error handling help us catch issues earlier.

jslivka commented on realtime-reporter/pkg/polaris/polaris.go:
```diff
@@ -0,0 +1,84 @@
+var config polarisconfiguration.Configuration
+
+func init() {
+	configPath := "examples/polaris.yaml"
```
i noticed this was still hardcoded, going to change this to a CLI flag
